File: java/src/main/java/home/playground/resources/TweetResource.java
package home.playground.resources;

import com.google.common.collect.Maps;
import com.google.common.collect.Sets;
import home.playground.models.SentimentAnalytics;
import home.playground.models.TweetAnalytics;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.function.Function;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseBody;

import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import java.sql.Timestamp;
import java.util.AbstractMap;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

/**
 * Created by cuong on 11/22/15.
 */
@Controller
public class TweetResource
{
    private static final Logger LOGGER = LoggerFactory.getLogger(TweetResource.class);

    @Autowired
    private JavaRDD<TweetAnalytics> tweetRdd;

    @RequestMapping(value="/")
    public String dashboard()
    {
        LOGGER.info("Get dashboard page");

        return "views/dashboard/page.html";
    }

    @ResponseBody
    @RequestMapping(value="/analyze/{hashTag}")
    @Produces(MediaType.APPLICATION_JSON)
    public List<SentimentAnalytics> analyze(@PathVariable("hashTag") String hashTag)
    {
        LOGGER.info("Sentiment analysis of " + hashTag);
        final Function<TweetAnalytics, Boolean> byHashTag =
            t -> t.getHashTag().equals(hashTag);

        return analyzeTweetAnalyticsRDD(tweetRdd.filter(byHashTag));
    }

    @ResponseBody
    @RequestMapping(value="/analyze")
    @Produces(MediaType.APPLICATION_JSON)
    public List<SentimentAnalytics> analyze(@RequestParam("hashTag")String hashTag, @RequestParam("start")Long start, @RequestParam("end")Long end)
    {
        LOGGER.info("Sentiment analysis of " + hashTag + " between " + start + " and " + end);

        final Function<TweetAnalytics, Boolean> byHashTag =
            t -> t.getHashTag().equals(hashTag);

        final Function<TweetAnalytics, Boolean> byCreatedDate =
            t -> 0 < t.getCreatedDate().compareTo(new Timestamp(start)) && t.getCreatedDate().compareTo(new Timestamp(end)) < 0;

        return analyzeTweetAnalyticsRDD(
            tweetRdd.filter(t -> byHashTag.call(t) && byCreatedDate.call(t))
        );
    }

    private List<SentimentAnalytics> analyzeTweetAnalyticsRDD(JavaRDD<TweetAnalytics> tweetAnalyticsRDD)
    {
        return tweetAnalyticsRDD.groupBy(TweetAnalytics::getSentiment)
            .mapValues(values -> {
                final Map<String, Map<String, Integer>> ageGroups = Maps.newHashMap(Maps.asMap(
                    Sets.newHashSet("child", "adult18to24", "adult25to34", "adult35to44", "adult45to54", "adult55to64", "adultOver64"),
                    ageGroup -> Maps.newHashMap(Maps.asMap(Sets.newHashSet("male", "female"), gender -> 0))
                ));

                return StreamSupport.stream(values.spliterator(), false)
                    .map(tweetAnalytics -> new AbstractMap.SimpleEntry<>(TweetAnalytics.resolveAgeGroup(tweetAnalytics.getAge()), tweetAnalytics.getGender()))
                    .reduce(
                        ageGroups,
                        (group, pair) -> {
                            group.get(pair.getKey()).compute(pair.getValue(), (gender, count) -> count + 1);

                            return group;
                        },
                        (x, acc) -> acc
                    );
            })
            .map (tuple -> {
                final SentimentAnalytics sentimentAnalytics = new SentimentAnalytics();
                sentimentAnalytics.setSentiment(tuple._1());
                sentimentAnalytics.setFreq(tuple._2());
                sentimentAnalytics.setTotal(tuple._2().entrySet().stream().flatMap(entry -> entry.getValue().values().stream()).reduce(0, (x, y) -> x + y));

                return sentimentAnalytics;
            })
            .collect();
    }

    @ResponseBody
    @RequestMapping(value="/timeline/{hashTag}")
    public  Map<String, List<List<Long>>> hist(@PathVariable("hashTag") String hashTag)
    {
        LOGGER.info("Sentiment timeline of " + hashTag);

        return tweetRdd.filter(it -> it.getHashTag().equals(hashTag))
            .groupBy(TweetAnalytics::getSentiment)
            .mapValues(
                it -> StreamSupport.stream(it.spliterator(), false)
                    .collect(Collectors.groupingBy(TweetAnalytics::getCreatedDate, Collectors.counting()))
                    .entrySet()
                    .stream()
                    .map(entry -> Arrays.asList(entry.getKey().getTime(), entry.getValue()))
                    .collect(Collectors.toList())
            )
            .collectAsMap();
    }
}

File: src/main/kotlin/home/playground/resources/TweetResource.kt
package home.playground.resources

import home.playground.models.*
import home.playground.services.FaceAnalysisService
import org.apache.spark.api.java.JavaSparkContext
import org.apache.spark.sql.SQLContext
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.ResponseBody
import java.util.*
import javax.ws.rs.Produces
import javax.ws.rs.core.MediaType


/**
 * Created by cuong on 11/14/15.
 */
@Controller
class TweetResource
{
    companion object {
        private val LOGGER = LoggerFactory.getLogger(TweetResource::class.java)
    }

    @Autowired
    private var javaSparkContext: JavaSparkContext? = null

    @Autowired
    private var sqlContext: SQLContext? = null

    private val fas = FaceAnalysisService()

    @RequestMapping(value="/")
    public fun dashboard(): String?
    {
        LOGGER.info("Get dashboard page")

        return "dashboard/page"
    }

    @ResponseBody
    @RequestMapping(value="/top")
    @Produces(MediaType.APPLICATION_JSON)
    public fun top(): List<HashTagStatistic>
    {
        LOGGER.info("Return top 10 hashtags")

        val hashTags = sqlContext!!.sql("SELECT hashTag, count(*) as count FROM tweets GROUP BY hashTag ORDER BY count desc LIMIT 10")

        return hashTags.javaRDD().collect().map { HashTagStatistic(hashTag = it.getString(0), count = it.getLong(1)) }
    }

    @ResponseBody
    @RequestMapping(value="/analyze/{hashTag}")
    @Produces(MediaType.APPLICATION_JSON)
    public fun analyze(@PathVariable("hashTag") hashTag: String): List<SentimentStatistic>
    {
        LOGGER.info("Sentiment analysis of $hashTag")

        val sentHist = HashMap<String, SentimentStatistic>()

        sqlContext!!.sql("SELECT sentiment, gender, age, count(*) as count FROM tweets WHERE hashTag = '$hashTag' GROUP BY sentiment, gender, age")
            .javaRDD().collect().forEach {
                val sent     = it.getString(0)
                val gender   = it.getString(1)
                val ageGroup = resolveAgeGroup(it.getInt(2))
                val count    = it.getLong(3)

                if (!sentHist.containsKey(sent)) {
                    sentHist.put(sent, SentimentStatistic(sentiment = sent))
                }

                val sentEntry = sentHist[sent] as SentimentStatistic
                val ageGroupEntry = sentEntry.freq.getRaw(ageGroup) as HashMap<String, Long>

                sentEntry.total += count
                ageGroupEntry.put(gender, ageGroupEntry[gender] as Long + count)
            }

        return sentHist.values.toList()
    }

    @ResponseBody
    @RequestMapping(value="/recognize")
    @Produces(MediaType.APPLICATION_JSON)
    public fun recognize(@RequestParam("imgUrl") imgUrl: String): FaceRecognization
    {
        /**
         * {
         *  positive: Array[Array[2], ..., Array[2]],
         *  negative: {...}
         * }
         */
        return fas.faceAnalysis(imgUrl)
    }

    private fun resolveAgeGroup(age: Int): String
    {
        if (age < 18) {
            return "child"
        } else if (age <= 24) {
            return "adult18to24"
        } else if (age <= 34) {
            return "adult25to34"
        } else if (age <= 44) {
            return "adult35to44"
        } else if (age <= 54) {
            return "adult45to54"
        } else if (age <= 64) {
            return "adult55to64"
        } else {
            return "adultOver64"
        }
    }
}